com.unity.multiplayer.mlapi/Runtime/Core/NetworkObject.cs

```diff
@@ -598,5 +599,169 @@ internal NetworkBehaviour GetNetworkBehaviourAtOrderIndex(ushort index)
+
+            NetworkObject parentNetworkObject = null;
+
+            if (!AlwaysReplicateAsRoot && transform.parent != null)
+            {
+                parentNetworkObject = transform.parent.GetComponent<NetworkObject>();
+            }
+
+            if (parentNetworkObject == null)
+            {
+                // We don't have a parent
+                writer.WriteBool(false);
+            }
+            else
+            {
+                // We do have a parent
+                writer.WriteBool(true);
+                // Write the parent's NetworkObjectId to be used for linking back to the child
+                writer.WriteUInt64Packed(parentNetworkObject.NetworkObjectId);
+            }
+
+            // Write if we are a scene object or not
+            writer.WriteBool(IsSceneObject ?? true);
+
+            // Write the hash for this NetworkObject
+            writer.WriteUInt32Packed(GlobalObjectIdHash);
+
+            if (IncludeTransformWhenSpawning == null || IncludeTransformWhenSpawning(OwnerClientId))
+            {
+                // Set the position and rotation data marker to true (i.e. flag to know, when reading from the stream, that position and rotation data follows).
+                writer.WriteBool(true);
+
+                // Write position
+                writer.WriteSinglePacked(transform.position.x);
+                writer.WriteSinglePacked(transform.position.y);
+                writer.WriteSinglePacked(transform.position.z);
+
+                // Write rotation
+                writer.WriteSinglePacked(transform.rotation.eulerAngles.x);
+                writer.WriteSinglePacked(transform.rotation.eulerAngles.y);
+                writer.WriteSinglePacked(transform.rotation.eulerAngles.z);
+            }
+            else
+            {
+                // Set the position and rotation data marker to false (i.e. flag to know, when reading from the stream, that position and rotation data *was not included*)
+                writer.WriteBool(false);
+            }
+
+            //Write whether we are including network variable data
+            writer.WriteBool(NetworkManager.NetworkConfig.EnableNetworkVariable);
+
+            if (NetworkManager.NetworkConfig.EnableNetworkVariable)
+            {
+                var buffer = writer.GetStream() as NetworkBuffer;
+                // Mark our current position and set the network variable data size initially to zero (NOT as a packed value)
+                var currentStreamPosition = buffer.Position;
+
+                // Write variable data size placeholder initially as zero
+                writer.WriteUInt32(0);
+
+                //Write the network variable data
+                WriteNetworkVariableData(writer.GetStream(), targetClientId);
+
+                //As long as something was written
+                if (buffer.Position > currentStreamPosition)
+                {
+                    //Get the new updated stream buffer position
```

Comment:
This code wasn't here before, seems a bit in an addition to the refactor.  Is this just an optimization to not send data if there's nothing to send?  Could there be a comment on what this overall is doing?

Reply:
**The entire block of code associated with this:**
```
 //Write whether we are including network variable data
  writer.WriteBool(NetworkManager.NetworkConfig.EnableNetworkVariable);
  
if (NetworkManager.NetworkConfig.EnableNetworkVariable)
{
    var buffer = writer.GetStream() as NetworkBuffer;
    // Mark our current position and set the network variable data size initially to zero (NOT as a packed value)
    var currentStreamPosition = buffer.Position;

    // Write variable data size placeholder initially as zero
    writer.WriteUInt32(0);

    //Write the network variable data
    WriteNetworkVariableData(writer.GetStream(), targetClientId);

    //As long as something was written
    if (buffer.Position > currentStreamPosition)
    {
        //Get the new updated stream buffer position
        var updatedPosition = buffer.Position;

        //Move back to the position just before we wrote our size
        buffer.Position = currentStreamPosition;

        // Replace the zero value placeholder with the size of data written into the stream by WriteNetworkVariableData
        writer.WriteUInt32((uint)(buffer.Position - currentStreamPosition));

        //Revert the buffer position back to the end of the network variable data written
        buffer.Position = updatedPosition;
    }
}
```
Basically, if EnableNetworkVariable is true then we will be including the size of the NetworkVariable data written into the stream for the NetworkObject in question.  We basically write a place holder (unpacked) UINT32 value, mark the position of the stream, then invoke WriteNetworkVariableData, and if the stream position changed we want to write the delta between the position prior to calling WriteNetworkVariableData and the current stream position.  This delta we will then write into the place holder position (i.e. before the WriteNetworkVariableData was invoked) so upon de-serializing this entry we will know the exact size of the Network Variable data in the event the NetworkObject in question fails to be instantiated on the de-serialization end of things (typically the client).

This is the "improved handling" of scene object serialization such that if one NetworkObject fails to be instantiated we can simply "log a warning" and skip to the next NetworkObject in the scene.

I think I provide enough comments for this, but if you have any additional suggestions on how to better improve the comments I am certainly open to suggestions.

Reply:
Ok, so it sounds like as part of the improved scene object handling, now when we provide the serialization for the object we also include the state of the its associated network variables.  I assume this is to avoid timing problems such as we talked about previously?

If so, great; I'd put in a comment to the effect of "force synchronize the state of this object's variables immediately to avoid consistency errors, etc."

Lastly, should make sure @jeffreyrainy sees this for context when we go to snapshot sync

Reply:
This only assists in the prevents completely asserting out of the current call-stack scope during scene synchronization.  This would happen during the InternalMessageHandler.HandleConnectionApproved and during the NetworkSceneManager.OnSceneUnloadClient.  The issue was that we couldn't just "skip an entry" in the serialized stream because we didn't know the size of the NetworkVariable data that was written into the stream for the NetworkObject in question that failed to instantiate on the client side.